keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,448 @@
+* `kube_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time.
+  * The value is the quantity of the request in `spec.containers.resources.limits`.
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `unit` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+
+The following metrics will be reported for pods that are in the `Running` phase and follow the same rules as above, but are not reported when the pod is `Pending` and can be used to calculate live resources. They all share the form `kube_running_pod_resource_(requests|limits)[_<resource>_<unit>]`.
+
+* `kube_running_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+* `kube_running_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time.
+
+
+This will add `O(pods * resource types)` metrics series, and in general most clusters have 4-8 resource types. The series per pod is `non_zero_resource_count * 2 * 2 (requests and limits for normal and _lifecycle)`.  This explicitly does not expose containers as a dimension, as the resource model applies rules to containers that cannot be trivially expressed via a formula and does not help answer the core questions.
+
+These metrics should roughly correspond to:
+
+* The resources the scheduler will consider the pod to be consuming at the current time
+* The resources the kubelet will use to decide whether to admit the pod (nodes will reject requested resources that exceed their actual allocatable capacity)
+* The effective reserved capacity as resource quota would observe for the namespace
+* The sum of all resources consumed on the cluster
+* The size of cgroup constraints applied to pods on a node
+
+The lifecycle metrics allow key insights that are impossible today:
+
+* Observing when init containers or containers have substantially different resource reservation, causing workloads to under provision nodes
+* Visualize batch compute consumption on the cluster where completed pods can be quickly excluded
+
+The implementation would be disablable via a flag in the kube-scheduler `--disable-pod-resource-metrics` to allow distributions to opt-out to the increased cardinality, although in practice we expect the impact in scheduler to be drowned by other O(pods) cardinality in other parts of the system and so this flag should not be necessary.
```

Comment:
So long as this is a separate endpoint, and no work is done unless the endpoint is queried, I don't think we need the flag.

Reply:
Hrm, did I not remove this section?  Will do so.